File: auth_samples_java/src/main/java/com/tuurio/authsample/JwtUtils.java
package com.tuurio.authsample;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;

public final class JwtUtils {
  private static final ObjectMapper MAPPER = new ObjectMapper();

  private JwtUtils() {}

  public static String decode(String token) {
    if (token == null || token.isEmpty()) {
      return "Not a JWT or unable to decode.";
    }
    String[] parts = token.split("\\.");
    if (parts.length < 2) {
      return "Not a JWT or unable to decode.";
    }
    try {
      byte[] decoded = Base64.getUrlDecoder().decode(padBase64(parts[1]));
      return prettyJson(new String(decoded, StandardCharsets.UTF_8));
    } catch (Exception ex) {
      return "Not a JWT or unable to decode.";
    }
  }

  public static String prettyJson(String raw) {
    try {
      Object json = MAPPER.readValue(raw, Object.class);
      return MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(json);
    } catch (Exception ex) {
      return raw;
    }
  }

  public static String prettyJson(Map<?, ?> claims) {
    try {
      return MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(claims);
    } catch (Exception ex) {
      return "No profile data.";
    }
  }

  private static String padBase64(String value) {
    int padding = 4 - (value.length() % 4);
    if (padding == 4) {
      return value;
    }
    return value + "=".repeat(padding);
  }
}

File: auth_samples_android/app/src/main/java/com/tuurio/authsample/auth/JwtUtils.kt
package com.tuurio.authsample.auth

import android.util.Base64
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.text.DateFormat
import java.util.Date

fun decodeJwt(token: String?): String? {
  if (token.isNullOrBlank()) return null
  val parts = token.split(".")
  if (parts.size < 2) return null

  return try {
    val decoded = Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    val json = String(decoded, StandardCharsets.UTF_8)
    JSONObject(json).toString(2)
  } catch (_: Exception) {
    null
  }
}

fun formatTime(epochMillis: Long?): String {
  if (epochMillis == null || epochMillis <= 0L) return "unknown time"
  val formatter = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.SHORT)
  return formatter.format(Date(epochMillis))
}
